src/catalog/src/stash.rs

```diff
@@ -50,148 +52,299 @@ use crate::transaction::{
+                Ok(stash) => stash,
+                Err(e) if e.can_recover_with_write_mode() => return Ok(false),
+                Err(e) => return Err(e.into()),
+            },
+            Some(stash) => stash,
+        };
+        stash.is_initialized().await.err_into()
+    }
 
-            let mut conn = Connection { stash };
+    async fn get_deployment_generation(&mut self) -> Result<Option<u64>, Error> {
+        let stash = match &mut self.stash {
+            None => match self.open_stash_read_only().await {
+                Ok(stash) => stash,
+                Err(e) if e.can_recover_with_write_mode() => return Ok(None),
+                Err(e) => return Err(e.into()),
+            },
+            Some(stash) => stash,
+        };
 
-            if !conn.stash.is_readonly() {
-                if let Some(deploy_generation) = deploy_generation {
-                    match conn.set_deploy_generation(deploy_generation).await {
-                        Ok(()) => {}
-                        Err(e) => {
-                            return Err((conn.stash, e));
-                        }
-                    }
+        let deployment_generation = CONFIG_COLLECTION
+            .peek_key_one(
+                stash,
+                proto::ConfigKey {
+                    key: DEPLOY_GENERATION.into(),
+                },
+            )
+            .await?
+            .map(|v| v.value);
+
+        Ok(deployment_generation)
+    }
+}
+
+/// Opens a new [`Connection`] to the stash, retrying any retryable errors. Optionally
+/// initialize the stash if it has not been initialized and perform any migrations needed.
+///
+/// # Panics
+/// If the inner stash has not been opened.
+async fn retry_open(
+    mut stash: Stash,
+    now: NowFn,
+    bootstrap_args: &BootstrapArgs,
+    deploy_generation: Option<u64>,
+) -> Result<Connection, Error> {
+    let retry = Retry::default()
+        .clamp_backoff(Duration::from_secs(1))
+        .max_duration(Duration::from_secs(30))
+        .into_retry_stream();
+    let mut retry = pin::pin!(retry);
+
+    loop {
+        match open_inner(stash, now.clone(), bootstrap_args, deploy_generation).await {
+            Ok(conn) => {
+                return Ok(conn);
+            }
+            Err((given_stash, err)) => {
+                stash = given_stash;
+                let should_retry = matches!(&err, Error::Stash(se) if se.should_retry());
+                if !should_retry || retry.next().await.is_none() {
+                    return Err(err);
                 }
             }
+        }
+    }
+}
 
-            conn
+#[tracing::instrument(name = "storage::open_inner", level = "info", skip_all)]
+async fn open_inner(
+    mut stash: Stash,
+    now: NowFn,
+    bootstrap_args: &BootstrapArgs,
+    deploy_generation: Option<u64>,
+) -> Result<Connection, (Stash, Error)> {
+    // Initialize the Stash if it hasn't been already
+    let mut conn = if !match stash.is_initialized().await {
```

Comment:
small thing, it might be nice to re-write this using `let-else` to avoid the "right drift", e.g.

```
let Ok(is_init) = stash.is_initialized().await else {
  return Err((stash, e.into());
};
```

You can probably also do this for all of the match statements in this function. It's just code formatting though so absolutely no need to do it if you don't want to!